cmd/proxy-to-grafana/proxy-to-grafana.go

```diff
@@ -0,0 +1,121 @@
+// proxy-to-grafana uses a reverse proxy which identifies users based on their originating tailscale node and creates / logs in users to grafana. Uses the auth proxy feature for grafana: https://grafana.com/docs/grafana/latest/auth/auth-proxy/
+//
+// Provide the env var TS_AUTHKEY to have this server automatically join your Tailnet.
+//
+// Use this grafana configuration to enable the auth proxy:
+
+// ```
+// [auth.proxy]
+// enabled = true
+// header_name = X-WEBAUTH-USER
+// header_property = username
+// auto_sign_up = true
+// whitelist = 127.0.0.1
+// headers = Name:X-WEBAUTH-NAME
+// enable_login_token = true
+// ```
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+
+	"inet.af/netaddr"
+	"tailscale.com/client/tailscale"
+	"tailscale.com/net/tsaddr"
+	"tailscale.com/tailcfg"
+	"tailscale.com/tsnet"
+)
+
+var (
+	hostname     = flag.String("hostname", "", "Tailscale hostname to serve on")
+	upstreamAddr = flag.String("upstream-addr", "", "Address of the upstream Grafana server, in host:port format")
+	tailscaleDir = flag.String("state-dir", "", "Directory in which to store the Tailscale auth state")
+)
+
+func main() {
+	flag.Parse()
+	if *hostname == "" {
+		log.Fatal("missing --hostname")
+	}
+	if *upstreamAddr == "" {
+		log.Fatal("missing --upstream-addr")
+	}
+	if *tailscaleDir == "" {
+		log.Fatal("missing --state-dir")
+	}
+	ts := &tsnet.Server{
+		Dir:      *tailscaleDir,
+		Hostname: *hostname,
+	}
+
+	url, err := url.Parse(fmt.Sprintf("http://%s", *upstreamAddr))
+	if err != nil {
+		panic(err)
+	}
+
+	proxy := httputil.NewSingleHostReverseProxy(url)
+	originalDirector := proxy.Director
+	proxy.Director = func(req *http.Request) {
+		originalDirector(req)
+		modifyRequest(req)
+	}
+
+	ltsn, err := ts.Listen("tcp", ":80")
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("serving access to %s on port 80", *upstreamAddr)
+	log.Fatal(http.Serve(ltsn, http.HandlerFunc(ProxyRequestHandler(proxy))))
+}
+
+func ProxyRequestHandler(proxy *httputil.ReverseProxy) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		proxy.ServeHTTP(w, r)
+	}
+}
+
+func modifyRequest(req *http.Request) {
+	// with enable_login_token set to true, we get a cookie that handles
+	// auth for paths that are not /login
+	if req.URL.Path != "/login" {
+		return
+	}
+
+	ipp, err := netaddr.ParseIPPort(req.RemoteAddr)
+	if err != nil {
+		log.Printf("bad remote address")
+		return
+	}
+	if !tsaddr.IsTailscaleIP(ipp.IP()) {
+		log.Printf("not a tailscale IP")
+		return
+	}
+	user, err := getTailscaleUser(req.Context(), req.RemoteAddr)
+	if err != nil {
+		log.Printf("error getting Tailscale user: %v", err)
+		return
+	}
+	// try to make these emails not collide with real users if any
+	email := strings.Replace(user.LoginName, "@", "-auto@", 1)
+	req.Header.Set("X-WEBAUTH-USER", email)
```

Comment:
X-Webauth-User
X-Webauth-Name

Reply:
Change made - but why is this more efficient?